packages/neo-one-client/src/user/LocalUserAccountProvider.ts

```diff
@@ -823,6 +841,172 @@ export class LocalUserAccountProvider<TKeyStore extends KeyStore, TProvider exte
     );
   }
 
+  private async consolidate({
+    inputs,
+    transactionUnsignedIn,
+    from,
+    monitor,
+    byteLimit,
+  }: {
+    readonly transactionUnsignedIn: TransactionModel;
+    readonly inputs: ReadonlyArray<UnspentOutput>;
+    readonly from: UserAccountID;
+    readonly monitor?: Monitor;
+    readonly byteLimit: number;
+  }): Promise<TransactionModel> {
+    const newInputSize = 68;
+    const newOutputSize = 120;
+    const messageSize = transactionUnsignedIn.serializeUnsigned().toString('hex').length;
+
+    if (byteLimit - messageSize < 0) {
+      throw new TransactionSizeExceedsLimitError(messageSize, byteLimit);
+    }
+    if (byteLimit - messageSize - newInputSize < 0) {
+      return transactionUnsignedIn;
+    }
+
+    const inputMap = (): { [key: string]: ReadonlyArray<UnspentOutput> } => ({});
+    const inputArray = (): ReadonlyArray<InputModel> => [];
+    const outputArray = (): ReadonlyArray<OutputModel> => [];
+    const getMessageSize = (numInputs: number, numOutputs: number) =>
+      messageSize + numInputs * newInputSize + numOutputs * newOutputSize;
+
+    const updateOutput = (output: OutputModel, amount: BigNumber): OutputModel =>
+      new OutputModel({
+        asset: output.asset,
+        value: output.value.add(clientUtils.bigNumberToBN(amount, 8)),
+        address: output.address,
+      });
+
+    const consoladatableUnspentsIn = await this.provider.getUnspentOutputs(from.network, from.address, monitor);
+
+    const looseAssets = consoladatableUnspentsIn
+      .filter(
+        (unspent) =>
+          !inputs.some((input) => unspent.txid === input.txid && unspent.vout === input.vout) &&
+          !this.mutableUsedOutputs.has(`${unspent.txid}:${unspent.vout}`),
+      )
+      .reduce((acc, unspent) => {
+        if ((acc[unspent.asset] as ReadonlyArray<UnspentOutput> | undefined) !== undefined) {
+          return {
+            ...acc,
+            [unspent.asset]: acc[unspent.asset].concat([unspent]),
+          };
+        }
+
+        return {
+          ...acc,
+          [unspent.asset]: [unspent],
+        };
+      }, inputMap());
+
+    const { newInputs, updatedOutputs, looserAssets } = transactionUnsignedIn.outputs
+      .filter(
+        (output) =>
+          (looseAssets[common.uInt256ToString(output.asset)] as ReadonlyArray<UnspentOutput> | undefined) !==
+            undefined && common.uInt160ToString(output.address) === addressToScriptHash(from.address),
+      )
+      .reduce(
+        (acc, output) => {
+          const currentAsset = common.uInt256ToString(output.asset);
+          const toAdd = Math.min(
+            looseAssets[currentAsset].length,
+            Math.floor((byteLimit - getMessageSize(acc.newInputs.length, 0)) / newInputSize),
+          );
+          if (toAdd === 0) {
+            return acc;
+          }
+          const tempIns = acc.looserAssets[currentAsset]
+            .sort((coinA, coinB) => coinA.value.comparedTo(coinB.value))
```

Comment:
IIRC this sorts in ascending order, is that the order we want? There should be a test case for this.

Reply:
Also sorting on every iteration seems odd too. `O(n^2logn)` complexity. It doesn't really matter in this case, but generally you should be trying to write optimal code so that it's structured consistently. Instead, just sort above (see comment).